File: Aircraft.cpp
#include "Aircraft.h"
#include "Utils.h"
#include <fstream>
#include <iostream>

int Aircraft::totalAircraftCount = 0;

Aircraft::Aircraft(const std::string& tn, const std::string& mdl, int total, int biz, int eco)
    : tailNumber(tn), model(mdl), totalSeats(total), businessSeats(biz), economySeats(eco),
    status(AircraftStatus::InService), flightHours(0.0), maintenanceSchedule("") {
    validate();
    ++totalAircraftCount;
}

void Aircraft::validate() {
    if (tailNumber.empty()) {
        throw std::invalid_argument("Бортовой номер не может быть пустым.");
    }
    if (model.empty()) {
        throw std::invalid_argument("Модель не может быть пустой.");
    }
    if (totalSeats <= 0 || businessSeats < 0 || economySeats < 0) {
        throw std::invalid_argument("Количество мест должно быть положительным.");
    }
    if (businessSeats + economySeats != totalSeats) {
        throw std::invalid_argument("Сумма мест бизнеса и эконома должна равняться общему.");
    }
}

std::string Aircraft::statusToString(AircraftStatus s) {
    switch (s) {
    case AircraftStatus::InService: return "В эксплуатации";
    case AircraftStatus::Maintenance: return "На ТО";
    case AircraftStatus::Retired: return "Списан";
    default: return "В эксплуатации";
    }
}

AircraftStatus Aircraft::stringToStatus(const std::string& s) {
    if (s == "В эксплуатации" || s == "InService") return AircraftStatus::InService;
    if (s == "На ТО" || s == "Maintenance") return AircraftStatus::Maintenance;
    if (s == "Списан" || s == "Retired") return AircraftStatus::Retired;
    return AircraftStatus::InService;
}

void Aircraft::scheduleMaintenance(const std::string& date) {
    if (date.length() != 10 || date[4] != '-' || date[7] != '-') {
        throw std::invalid_argument("Дата ТО должна быть в формате ГГГГ-ММ-ДД.");
    }
    maintenanceSchedule = date;
    changeStatus(AircraftStatus::Maintenance);
}

void Aircraft::updateFlightHours(double hours) {
    if (hours < 0) {
        throw std::invalid_argument("Налёт не может быть отрицательным.");
    }
    flightHours += hours;
}

void Aircraft::changeStatus(AircraftStatus st) {
    status = st;
    if (this->status == AircraftStatus::Retired) {
        this->flightHours = 0.0;
    }
}

bool Aircraft::operator==(const Aircraft& other) const {
    return this->tailNumber == other.tailNumber;
}

bool Aircraft::operator!=(const Aircraft& other) const {
    return !(*this == other);
}

std::ostream& operator<<(std::ostream& os, const Aircraft& ac) {
    os << "Самолёт[Борт=" << ac.tailNumber << ", Модель=" << ac.model
        << ", Статус=" << Aircraft::statusToString(ac.status)
        << ", Налёт=" << ac.flightHours << "]";
    return os;
}

void Aircraft::loadAll(std::vector<std::unique_ptr<Aircraft>>& aircrafts) {
    aircrafts.clear();
    std::ifstream file("data/aircrafts.txt");
    if (!file.is_open()) return;

    std::string line;
    while (std::getline(file, line)) {
        line.pop_back();
        if (line.empty() || line[0] == '#') continue;
        auto parts = split(line, ';');
        if (parts.size() != 8) continue;

        try {
            auto ac = std::make_unique<Aircraft>();
            ac->tailNumber = parts[0];
            ac->model = parts[1];
            ac->totalSeats = std::stoi(parts[2]);
            ac->businessSeats = std::stoi(parts[3]);
            ac->economySeats = std::stoi(parts[4]);
            ac->status = stringToStatus(parts[5]);
            ac->maintenanceSchedule = parts[6];
            ac->flightHours = std::stod(parts[7]);
            ac->validate();
            aircrafts.push_back(std::move(ac));
        }
        catch (const std::exception& e) {
            std::cerr << "Пропущен некорректный самолёт:" << e.what() << "\n";
        }
    }
    file.close();
}

void Aircraft::saveAll(const std::vector<std::unique_ptr<Aircraft>>& aircrafts) {
    ensureDataDir();
    std::ofstream file("data/aircrafts.txt");
    file << "# Tail;Model;Total;Business;Economy;Status;Maintenance;FlightHours\n";
    for (const auto& ac : aircrafts) {
        file << ac->tailNumber << ";"
            << ac->model << ";"
            << ac->totalSeats << ";"
            << ac->businessSeats << ";"
            << ac->economySeats << ";"
            << (ac->status == AircraftStatus::InService ? "InService" :
                ac->status == AircraftStatus::Maintenance ? "Maintenance" : "Retired") << ";"
            << ac->maintenanceSchedule << ";"
            << ac->flightHours << "\n";
    }
    file.close();
}

File: Aircraft.cs
amespace Lab4
{
    public enum AircraftStatus
{
    InService,
	@@ -14,13 +13,16 @@ public enum AircraftStatus

    public class Aircraft
{
    // Статическое поле — для подсчёта общего числа ВС
    public static int TotalCount { get; private set; } = 0;

    public string TailNumber { get; set; }
    public string Model { get; set; }
    public int TotalSeats { get; set; }
    public int BusinessSeats { get; set; }
    public int EconomySeats { get; set; }
    public AircraftStatus Status { get; set; }
    public string MaintenanceSchedule { get; set; }
    public double FlightHours { get; set; }

    public Aircraft() { }
public Aircraft(string tailNumber, string model, int totalSeats, int businessSea
            EconomySeats = economySeats;
    Status = AircraftStatus.InService;
            FlightHours = 0.0;
            TotalCount++; // инкремент статического счётчика
        }

public static string StatusToString(AircraftStatus status)
	 public static List<Aircraft> LoadAll()
            foreach (var line in File.ReadAllLines("data/aircrafts.txt"))
{
    if (string.IsNullOrWhiteSpace(line) || line.StartsWith("#")) continue;
    var parts = Utils.Split(line, ';');
    if (parts.Count != 8) continue;

	public static List<Aircraft> LoadAll()
                    MaintenanceSchedule = parts[6],
                    FlightHours = double.Parse(parts[7])
                };
                // TotalCount не инкрементируем здесь — только в конструкторе при создании новых
            }
            return aircrafts;
        }

        private static string LeftPad(string text, int width)
{
    if (text == null) text = "";
    return text.Length > width ? text.Substring(0, width) : text.PadRight(width);
}

public static void SaveAll(List<Aircraft> aircrafts)
{
    Utils.EnsureDataDir();
	@@ -117,9 + 118,10 @@ public static void SaveAll(List<Aircraft> aircrafts)
            foreach (var ac in aircrafts)
{
    writer.WriteLine($"{ac.TailNumber};{ac.Model};{ac.TotalSeats};{ac.BusinessSeats};{ac.EconomySeats};" +
$"{StatusToString(ac.Status)};{ac.MaintenanceSchedule};{ac.FlightHours}");
}
        }

        public static void DisplayTable(List<Aircraft> list)
{
    if (list == null || list.Count == 0)
	public static void DisplayTable(List<Aircraft> list)
            }

            Console.WriteLine("\n=== Воздушные суда ===");
int tailWidth = 12;
int modelWidth = 18;
int totalWidth = 8;
int bizWidth = 10;
int statusWidth = 14;
int hoursWidth = 12;

Console.WriteLine(
    $"{LeftPad("Борт", tailWidth)} " +
    $"{LeftPad("Модель", modelWidth)} " +
